From 3c9d34e3de1828f2133789048707ccbec1cf75f0 Mon Sep 17 00:00:00 2001
From: David Daney <david.daney@cavium.com>
Date: Thu, 16 May 2019 21:49:11 +0200
Subject: [PATCH 293/345] MIPS/tlbex: Save and restore ASID around TLBR

Not sure why this is needed, but fixes userspace corruption problems
seen by a customer.

Signed-off-by: David Daney <david.daney@cavium.com>
---
 arch/mips/mm/tlbex.c | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/arch/mips/mm/tlbex.c b/arch/mips/mm/tlbex.c
index 942e03bc1bd7..a2649065e363 100644
--- a/arch/mips/mm/tlbex.c
+++ b/arch/mips/mm/tlbex.c
@@ -2174,6 +2174,7 @@ static void build_r4000_tlb_load_handler(void)
 		 */
 		WARN(cpu_has_tlbex_tlbp_race(), "Unhandled race in RiXi path");
 
+		UASM_i_MFC0(&p, wr.r3, C0_ENTRYHI);
 		uasm_i_tlbr(&p);
 
 		switch (current_cpu_type()) {
@@ -2188,6 +2189,8 @@ static void build_r4000_tlb_load_handler(void)
 			}
 		}
 
+		UASM_i_MTC0(&p, wr.r3, C0_ENTRYHI);
+
 		/* Examine  entrylo 0 or 1 based on ptr. */
 		if (use_bbit_insns()) {
 			uasm_i_bbit0(&p, wr.r2, ilog2(sizeof(pte_t)), 8);
@@ -2249,6 +2252,7 @@ static void build_r4000_tlb_load_handler(void)
 		 */
 		WARN(cpu_has_tlbex_tlbp_race(), "Unhandled race in RiXi path");
 
+		UASM_i_MFC0(&p, wr.r3, C0_ENTRYHI);
 		uasm_i_tlbr(&p);
 
 		switch (current_cpu_type()) {
@@ -2263,6 +2267,8 @@ static void build_r4000_tlb_load_handler(void)
 			}
 		}
 
+		UASM_i_MTC0(&p, wr.r3, C0_ENTRYHI);
+
 		/* Examine  entrylo 0 or 1 based on ptr. */
 		if (use_bbit_insns()) {
 			uasm_i_bbit0(&p, wr.r2, ilog2(sizeof(pte_t)), 8);
-- 
2.25.1

